diff --git a/src/gallium/drivers/lima/lima_job.c b/src/gallium/drivers/lima/lima_job.c
index 3ec53db2d2a0255067ac6f0a745ca020266dd86f..14f52da4c7e96a047ecd5302cb789000f177bcd0 100644
--- a/src/gallium/drivers/lima/lima_job.c
+++ b/src/gallium/drivers/lima/lima_job.c
@@ -1112,6 +1112,14 @@ lima_pipe_flush(struct pipe_context *pctx, struct pipe_fence_handle **fence,
    }
 }
 
+static void
+lima_texture_barrier(struct pipe_context *pctx, unsigned flags)
+{
+    struct lima_context *ctx = lima_context(pctx);
+
+    lima_flush(ctx);
+}
+
 static bool
 lima_job_compare(const void *s1, const void *s2)
 {
@@ -1146,6 +1154,7 @@ bool lima_job_init(struct lima_context *ctx)
    }
 
    ctx->base.flush = lima_pipe_flush;
+   ctx->base.texture_barrier = lima_texture_barrier;
 
    return true;
 }
diff --git a/src/gallium/drivers/lima/lima_screen.c b/src/gallium/drivers/lima/lima_screen.c
index 6215b2bc7e961fef0906b9fd35ce29a2d1a3ee9d..a04cf0c6373017152839884508be2877caffcca2 100644
--- a/src/gallium/drivers/lima/lima_screen.c
+++ b/src/gallium/drivers/lima/lima_screen.c
@@ -106,6 +106,7 @@ lima_screen_get_param(struct pipe_screen *pscreen, enum pipe_cap param)
    case PIPE_CAP_FRAGMENT_SHADER_TEXTURE_LOD:
    case PIPE_CAP_TEXTURE_SWIZZLE:
    case PIPE_CAP_VERTEX_COLOR_UNCLAMPED:
+   case PIPE_CAP_TEXTURE_BARRIER:
       return 1;
 
    /* Unimplemented, but for exporting OpenGL 2.0 */
